File: backend/app/simulation/edit_policy.py
# -*- coding: utf-8 -*-
"""EditPolicy v1 (M7.14D) — affordance chỉnh sửa DẪN XUẤT TỪ NĂNG LỰC của cảnh.

Vấn đề đang sửa: mọi cảnh generic đều nhận CÙNG một bộ thao tác (Thêm điểm/Nối/
Xóa), kể cả cảnh văn bản (container/heading/paragraph) hay cảnh giá trị
(switch/lamp/value_box). Sai về ngữ nghĩa.

Nguyên tắc:
- Suy từ CHÍNH SPEC (object/rule/process types) — spec vẫn đúng SAU khi patch,
  khác với analysis của đề gốc (analysis không có ở sample offline, không cập
  nhật sau edit).
- TUYỆT ĐỐI không hard-code theo tên bài/môn/tiêu đề ("triangle", "web"...).
- Ẩn UI là KHÔNG đủ: policy được thực thi ở cả patch validation (file này được
  patch.py + edit.py gọi) lẫn affordance (mirror TS: generic/edit-policy.ts).

LƯU Ý PHẠM VI: `EditFamily` dưới đây là phân loại của EditPolicy **v1**, KHÔNG
phải taxonomy vĩnh viễn của hệ thống (taxonomy vĩnh viễn là SEMANTIC_ROLES trong
manifest). Cảnh LAI (vd vừa structural vừa node/edge) dùng precedence BẢO THỦ —
multi-family edit CHƯA được hỗ trợ, xem docs/ARCHITECTURE_MAP.md.
"""

from __future__ import annotations

from app.simulation.dsl.manifest import limit, temporal_process_types

# ── reason_code (M7.14D) — hai namespace ─────────────────────
# policy.*    : thao tác hợp lệ về cấu trúc nhưng KHÔNG hợp với năng lực cảnh này
# structure.* : vi phạm luật DSL (id trùng, tham chiếu treo, chu trình, quá giới hạn)
POLICY_OPERATION_NOT_ALLOWED = "policy.operation_not_allowed"
POLICY_OBJECT_TYPE_NOT_ALLOWED = "policy.object_type_not_allowed"
POLICY_PATH_TOPOLOGY_LOCKED = "policy.path_topology_locked"
POLICY_FAMILY_MISMATCH = "policy.family_mismatch"  # CHỈ dùng khi không có code cụ thể hơn

STRUCTURE_INVALID = "structure.invalid"  # mặc định cho lỗi DSL (validator quyết chi tiết)

# Họ object (mirror generic/model.ts)
STRUCTURAL_TYPES = {"container", "group", "heading", "paragraph", "text"}
TEXT_CONTENT_TYPES = {"heading", "paragraph", "text"}
CONTAINER_TYPES = {"container", "group"}
RELATIONAL_TYPES = {"node", "edge"}

ALL_OPS = ("add_object", "remove_object", "update_object", "connect", "disconnect")


class EditFamily:
    """Phân loại EditPolicy v1 (không phải taxonomy vĩnh viễn của hệ)."""

    SPATIAL = "spatial"            # điểm/nút + cạnh tường minh → thêm điểm/nối/xóa
    STRUCTURAL = "structural"      # khung chứa + nội dung chữ → thêm/sửa/xóa nội dung
    VALUE_ONLY = "value_only"      # switch/lamp/value_box + rule → chỉ tương tác, không sửa cấu trúc
    OBSERVATION = "observation"    # cảnh có tiến trình di chuyển → topology KHÓA


def _types_in(spec: dict) -> set[str]:
    return {o.get("type", "") for o in spec.get("objects", [])}


def _has_move_process(spec: dict) -> bool:
    """Process di chuyển ràng buộc topology (path/entity) → không cho sửa cấu trúc."""
    return any(p.get("type") == "move_along_path" for p in spec.get("processes", []))


def _max_depth(spec: dict) -> int:
    by_id = {o["id"]: o for o in spec.get("objects", [])}
    best = 0
    for o in spec.get("objects", []):
        depth, cur, seen = 0, o["id"], {o["id"]}
        while by_id.get(cur, {}).get("parent") is not None:
            cur = by_id[cur]["parent"]
            if cur in seen:
                break
            seen.add(cur)
            depth += 1
        best = max(best, depth)
    return best


def edit_policy_of(spec: dict) -> dict:
    """Suy EditPolicy từ spec. Precedence BẢO THỦ (cảnh lai → chọn hạn chế hơn):

        move_along_path  >  structural  >  spatial  >  value_only

    Trả: {family, allowed_ops, addable_types, ui_actions, note}
    """
    types = _types_in(spec)
    has_structural = bool(types & STRUCTURAL_TYPES)
    has_relational = bool(types & RELATIONAL_TYPES)

    # 1) Tiến trình di chuyển: thêm/xóa node đổi ngữ nghĩa path → KHÓA topology.
    if _has_move_process(spec):
        return {
            "family": EditFamily.OBSERVATION,
            "allowed_ops": ["update_object"],  # đổi nhãn/nội dung là vô hại
            "addable_types": [],
            "ui_actions": ["edit_text"],
            "note": "Cảnh có tiến trình di chuyển theo đường — cấu trúc topology bị khóa.",
        }

    # 2) Cảnh cấu trúc/nội dung: thêm mục nội dung, KHÔNG thêm điểm/nối.
    if has_structural:
        addable = sorted(TEXT_CONTENT_TYPES)
        # container/group chỉ thêm được khi còn dư độ sâu lồng nhau
        if _max_depth(spec) + 1 < limit("max_nesting_depth"):
            addable += sorted(CONTAINER_TYPES)
        return {
            "family": EditFamily.STRUCTURAL,
            "allowed_ops": ["add_object", "remove_object", "update_object"],
            "addable_types": addable,
            "ui_actions": ["add_content", "edit_text", "delete"],
            "note": "Cảnh nội dung có bố cục — thêm/sửa/xóa mục nội dung.",
        }

    # 3) Cảnh quan hệ điểm-cạnh tường minh: đây (và CHỈ đây) mới có Thêm điểm/Nối.
    if has_relational:
        return {
            "family": EditFamily.SPATIAL,
            "allowed_ops": list(ALL_OPS),
            "addable_types": ["node", "edge", "label"],
            "ui_actions": ["add_node", "connect", "delete", "edit_text"],
            "note": "Cảnh điểm–cạnh — thêm điểm, nối, xóa.",
        }

    # 4) Còn lại (switch/lamp/value_box + rule): tương tác giữ nguyên, KHÔNG sửa
    #    cấu trúc (thêm/xóa object sẽ làm rule sai nghĩa).
    return {
        "family": EditFamily.VALUE_ONLY,
        "allowed_ops": ["update_object"],
        "addable_types": [],
        "ui_actions": ["edit_text"],
        "note": "Cảnh giá trị/logic — dùng tương tác sẵn có (bật/tắt), không sửa cấu trúc.",
    }


def check_ops_against_policy(spec: dict, ops: list[dict]) -> dict | None:
    """Kiểm ops theo policy của spec. Trả None nếu hợp lệ, hoặc
    {"reason_code", "error"} nếu vi phạm — CODE CỤ THỂ được ưu tiên,
    family_mismatch chỉ là fallback."""
    policy = edit_policy_of(spec)
    allowed_ops = set(policy["allowed_ops"])
    addable = set(policy["addable_types"])
    locked = policy["family"] == EditFamily.OBSERVATION

    for op in ops:
        kind = op.get("op")
        if kind not in allowed_ops:
            if locked and kind in ("add_object", "remove_object", "connect", "disconnect"):
                return {
                    "reason_code": POLICY_PATH_TOPOLOGY_LOCKED,
                    "error": (
                        f'Không thể "{kind}" trong cảnh này: có tiến trình di chuyển theo đường, '
                        "thay đổi topology sẽ làm sai đường đi. " + policy["note"]
                    ),
                }
            return {
                "reason_code": POLICY_OPERATION_NOT_ALLOWED,
                "error": f'Thao tác "{kind}" không phù hợp với cảnh này. {policy["note"]}',
            }
        if kind == "add_object":
            obj_type = (op.get("object") or {}).get("type")
            if obj_type not in addable:
                return {
                    "reason_code": POLICY_OBJECT_TYPE_NOT_ALLOWED,
                    "error": (
                        f'Không thể thêm đối tượng loại "{obj_type}" vào cảnh này. '
                        + (f'Chỉ thêm được: {", ".join(sorted(addable))}.' if addable
                           else "Cảnh này không cho thêm đối tượng mới.")
                    ),
                }
    return None


def policy_contract_text(spec: dict) -> str:
    """Mô tả policy cho prompt edit — LLM chỉ thấy thao tác/loại HỢP LỆ với cảnh
    này (prompt nhỏ hơn, ít patch sai hơn). Server vẫn chặn tất định sau đó."""
    p = edit_policy_of(spec)
    if not p["allowed_ops"] or p["allowed_ops"] == ["update_object"]:
        return (
            f"PHẠM VI CHỈNH SỬA CỦA CẢNH NÀY: {p['note']}\n"
            'Chỉ được dùng: update_object (đổi "text"/"label" của đối tượng có sẵn). '
            "Mọi yêu cầu thêm/xóa/nối đối tượng đều KHÔNG hợp lệ với cảnh này — "
            "khi đó trả operations rỗng và nêu lý do trong note."
        )
    return (
        f"PHẠM VI CHỈNH SỬA CỦA CẢNH NÀY: {p['note']}\n"
        f"Thao tác được phép: {', '.join(p['allowed_ops'])}.\n"
        f"Loại đối tượng được phép thêm: {', '.join(p['addable_types']) or 'không có'}.\n"
        "Yêu cầu nằm ngoài phạm vi này KHÔNG hợp lệ — trả operations rỗng và nêu lý do trong note."
    )

File: frontend/src/simulations/domains/generic/edit-policy.ts
import {
  CONTAINER_TYPES,
  STRUCTURAL_TYPES,
  TEXT_CONTENT_TYPES,
  type ObjectType,
  type SimulationSpec,
} from "./model";

/**
 * EditPolicy v1 (M7.14D) — mirror app/simulation/edit_policy.py.
 *
 * Affordance chỉnh sửa DẪN XUẤT TỪ NĂNG LỰC của cảnh, không phải mặc định cho
 * mọi cảnh generic. Suy từ CHÍNH SPEC (object/process types) — spec vẫn đúng
 * sau khi patch, khác với analysis của đề gốc (không có ở sample offline).
 * TUYỆT ĐỐI không hard-code theo tên bài/môn/tiêu đề.
 *
 * PHẠM VI: `EditFamily` là phân loại của EditPolicy **v1**, KHÔNG phải taxonomy
 * vĩnh viễn của hệ (taxonomy vĩnh viễn là SEMANTIC_ROLES trong manifest).
 * Cảnh LAI dùng precedence bảo thủ — multi-family edit CHƯA hỗ trợ.
 */

export type EditFamily = "spatial" | "structural" | "value_only" | "observation";

export type PatchOpKind = "add_object" | "remove_object" | "update_object" | "connect" | "disconnect";

/** Thao tác UI (không phải patch op): quyết định nút nào hiện ra. */
export type EditUiAction = "add_node" | "add_content" | "connect" | "delete" | "edit_text";

export interface EditPolicy {
  family: EditFamily;
  allowedOps: PatchOpKind[];
  addableTypes: ObjectType[];
  uiActions: EditUiAction[];
  note: string;
}

/** reason_code — hai namespace, mirror backend. */
export const POLICY_OPERATION_NOT_ALLOWED = "policy.operation_not_allowed";
export const POLICY_OBJECT_TYPE_NOT_ALLOWED = "policy.object_type_not_allowed";
export const POLICY_PATH_TOPOLOGY_LOCKED = "policy.path_topology_locked";
export const POLICY_FAMILY_MISMATCH = "policy.family_mismatch"; // fallback, ưu tiên code cụ thể
export const STRUCTURE_INVALID = "structure.invalid";

const RELATIONAL_TYPES = new Set<string>(["node", "edge"]);
const MAX_NESTING_DEPTH = 4; // mirror manifest limits.max_nesting_depth

function hasMoveProcess(spec: SimulationSpec): boolean {
  return spec.processes.some((p) => p.type === "move_along_path");
}

function maxDepth(spec: SimulationSpec): number {
  const byId = new Map(spec.objects.map((o) => [o.id, o]));
  let best = 0;
  for (const o of spec.objects) {
    let depth = 0;
    let cur: string | undefined = o.id;
    const seen = new Set<string>([o.id]);
    while (cur !== undefined && byId.get(cur)?.parent !== undefined) {
      cur = byId.get(cur)!.parent;
      if (cur === undefined || seen.has(cur)) break;
      seen.add(cur);
      depth += 1;
    }
    best = Math.max(best, depth);
  }
  return best;
}

/**
 * Precedence BẢO THỦ (cảnh lai → chọn family hạn chế hơn):
 *   move_along_path > structural > spatial > value_only
 */
export function editPolicyOf(spec: SimulationSpec): EditPolicy {
  const types = new Set<string>(spec.objects.map((o) => o.type));
  const hasStructural = [...types].some((t) => STRUCTURAL_TYPES.has(t));
  const hasRelational = [...types].some((t) => RELATIONAL_TYPES.has(t));

  // 1) Tiến trình di chuyển: thêm/xóa node đổi ngữ nghĩa path → KHÓA topology.
  if (hasMoveProcess(spec)) {
    return {
      family: "observation",
      allowedOps: ["update_object"],
      addableTypes: [],
      uiActions: ["edit_text"],
      note: "Cảnh có tiến trình di chuyển theo đường — cấu trúc topology bị khóa.",
    };
  }

  // 2) Cảnh cấu trúc/nội dung: thêm mục nội dung, KHÔNG thêm điểm/nối.
  if (hasStructural) {
    const addable = [...TEXT_CONTENT_TYPES].sort() as ObjectType[];
    if (maxDepth(spec) + 1 < MAX_NESTING_DEPTH) {
      addable.push(...([...CONTAINER_TYPES].sort() as ObjectType[]));
    }
    return {
      family: "structural",
      allowedOps: ["add_object", "remove_object", "update_object"],
      addableTypes: addable,
      uiActions: ["add_content", "edit_text", "delete"],
      note: "Cảnh nội dung có bố cục — thêm/sửa/xóa mục nội dung.",
    };
  }

  // 3) Cảnh quan hệ điểm–cạnh: CHỈ ở đây mới có Thêm điểm/Nối.
  if (hasRelational) {
    return {
      family: "spatial",
      allowedOps: ["add_object", "remove_object", "update_object", "connect", "disconnect"],
      addableTypes: ["node", "edge", "label"],
      uiActions: ["add_node", "connect", "delete", "edit_text"],
      note: "Cảnh điểm–cạnh — thêm điểm, nối, xóa.",
    };
  }

  // 4) Còn lại (switch/lamp/value_box + rule): tương tác giữ nguyên, không sửa cấu trúc.
  return {
    family: "value_only",
    allowedOps: ["update_object"],
    addableTypes: [],
    uiActions: ["edit_text"],
    note: "Cảnh giá trị/logic — dùng tương tác sẵn có (bật/tắt), không sửa cấu trúc.",
  };
}

/**
 * Có công cụ sửa nào ĐÁNG để mở chế độ Chỉnh sửa không? (M7.14D.1)
 *
 * Suy TỪ POLICY, không hard-code theo binary/logic/tiêu đề. `edit_text` một
 * mình KHÔNG tính là "đáng": nó không có công cụ trực quan nào trên sân khấu,
 * nên mở chế độ Chỉnh sửa chỉ để hiện một ô nhập trống là quảng bá một
 * affordance rỗng. Backend policy vẫn giữ nguyên (update_object qua NL edit vẫn
 * hợp lệ nếu ai đó gọi API) — đây chỉ là chuyện UI không mời gọi.
 */
export function hasMeaningfulEditAffordance(policy: EditPolicy): boolean {
  return policy.uiActions.some((a) => a !== "edit_text");
}

export interface PolicyViolation {
  reasonCode: string;
  error: string;
}

/** Kiểm ops theo policy. null = hợp lệ. Code CỤ THỂ được ưu tiên hơn family_mismatch. */
export function checkOpsAgainstPolicy(
  spec: SimulationSpec,
  ops: { op?: string; object?: { type?: string } }[],
): PolicyViolation | null {
  const policy = editPolicyOf(spec);
  const allowed = new Set<string>(policy.allowedOps);
  const addable = new Set<string>(policy.addableTypes);
  const locked = policy.family === "observation";

  for (const op of ops) {
    const kind = op?.op ?? "";
    if (!allowed.has(kind)) {
      if (locked && ["add_object", "remove_object", "connect", "disconnect"].includes(kind)) {
        return {
          reasonCode: POLICY_PATH_TOPOLOGY_LOCKED,
          error: `Không thể "${kind}" trong cảnh này: có tiến trình di chuyển theo đường, thay đổi topology sẽ làm sai đường đi. ${policy.note}`,
        };
      }
      return {
        reasonCode: POLICY_OPERATION_NOT_ALLOWED,
        error: `Thao tác "${kind}" không phù hợp với cảnh này. ${policy.note}`,
      };
    }
    if (kind === "add_object") {
      const objType = op.object?.type;
      if (!objType || !addable.has(objType)) {
        return {
          reasonCode: POLICY_OBJECT_TYPE_NOT_ALLOWED,
          error:
            `Không thể thêm đối tượng loại "${String(objType)}" vào cảnh này. ` +
            (addable.size > 0
              ? `Chỉ thêm được: ${[...addable].sort().join(", ")}.`
              : "Cảnh này không cho thêm đối tượng mới."),
        };
      }
    }
  }
  return null;
}

/** Nhãn tiếng Việt của loại nội dung thêm được (dropdown "Thêm nội dung"). */
export const ADDABLE_TYPE_LABEL: Record<string, string> = {
  heading: "Tiêu đề",
  paragraph: "Đoạn văn",
  text: "Dòng chữ",
  container: "Khung chứa",
  group: "Nhóm",
  node: "Điểm/nút",
  edge: "Cạnh",
  label: "Nhãn",
};
